Plugins

aschuch opened the pull request:
I would like to propose a plugin system for Moya.
`MoyaProvider` can now be initialized with a set of `Plugin`s.  A `Plugin` receives callbacks to perform side effects wherever a request is sent or received.

For example, a plugin may be used to
 * log network requests
 * hide and show a network avtivity indicator
 * inject additional information into a request

This kind of tackles #144 as well, as the number of additional closures is now much lower and can be initialized with an arbitrary number of plugins

I have also added plugins for the existing `CredentialClosure` and `NetworkActivityClosure` to make it easy for everyone to adopt this change. Maybe we should also try to build a plugin from the Logger in #235?

Would love to get some feedback on the idea or how to make this even more generic for everyone to use (I'm not super happy with the subclassing approach). :heart: 


Fixes #144.

aloco commented:
I really like this approach! Since Moya get´s more and more awesomeness, there should be a consistent way in adding new features without having dozens of specific closures (like the already mentioned credentials, logger, network activity). 
It would also reduce the amount of breaking changes if new features were implemented as ```Plugin```s whenever possible.

ashfurrow commented on Moya/Moya.swift:
```diff
@@ -212,30 +197,22 @@ public extension MoyaProvider {
-        if let cred = credential {
-            request = request.authenticate(usingCredential: cred)
+        // Give plugins the chance to alter the outgoing request
+        for p in plugins {
```
Maybe something like...

```swift
plugins.forEach{ $0.willSendRequest(token, request: request) }
```

Does that make sense?

aschuch replied:
Huh, I put that in for debugging. Good catch!

ashfurrow commented on Moya/Plugins/CredentialsPlugin.swift:
```diff
@@ -0,0 +1,31 @@
+    override func willSendStubbedRequest(token: Target, request: NSURLRequest) {
+        // Just call the closure here to make it possible to test the 
+        // credentials closure when stubbed requests are turned on
+        let _ = credentialsClosure(token)
```
You can drop the `let` here.

aschuch replied:
:tada:

ashfurrow commented on Moya/Plugin.swift:
```diff
@@ -0,0 +1,44 @@
+    // 
+    // This does not work, because `plugins` is now unable to infer the actual type of the typealias `T`.
+    
+    func willSendRequest(token: Target, request: Alamofire.Request) -> Alamofire.Request {
```
Why don't we overload this method with an identical one, except that it doesn't return a request? That way if the function doesn't need to modify the request, it can just omit `return` and let the type inference of Swift handle it?

aschuch replied:
Good point. Alamofire.Request is passed by reference anyway, maybe we can get rid of the return anyway?

In case we want to go forward with the overloading, calling just  `plugins.forEach{ request = $0.willSendRequest(token, request: request) }` would no longer work, since it will always call the function with the return value.

Should we invoke both methods within the `forEach` closure?

ashfurrow replied:
I hadn't thought of that, actually. Good point. If the request can be modified in-place, may as well get rid of the return value.

ashfurrow commented:
Generally looks great! I was a bit wary of the idea of "plugins", but I really like the implementation. 

I have a few pieces of general feedback:

- I saw the note about avoiding protocols – let's see if we can think of something besides inheritance? Nothing comes to mind, though ...
- Should a plugin be able to distinguish between stubbed requests and real ones? 
- Maybe the plugin functions should accept the `manager` as a parameter? I dunno, probably unnecessary but it feels like something that _should_ happen. Thoughts?
- Could we call them something other than "plugins"? :stuck_out_tongue: 

Thanks again!

![](http://media.giphy.com/media/V2xbsCrxcLQSQ/giphy.gif)

aschuch commented:
Thanks for the feedback. I'll push your remarks asap.

* First of, I don't really like the subclassing approach either. I tried a few things with protocols but none of them really seemed to work out. I'd love to discuss alternatives though.
* I decided to give the `Plugin` separate calls for stubbed requests and real ones, because the parameters are a bit different. This could, however, be abstracted into a single method on the plugin that accepts a stubbed/not-stubbed parameter.
* Could easily add the `manager` to the plugin functions, not sure what the plugin would do with it though. Do you have any use cases that come to mind?
* I am not too fond on the term _plugin_ either. Maybe _mixin_? _sideeffect_? I just named it plugin, because everyone is sort of familiar of this term. 

:boom:

ashfurrow commented:
Ahhh, sorry I meant the `provider` parameter (through which, I guess, the `manager` would be accessible too).

I guess what I'm _really_ saying about the stubbed/not stubbed functions is that I don't think the plugin should even know if the request it's being called with is being stubbed or not. 

Nomenclature is hard, let's give it a think :)

ashfurrow commented:
Also, I agree with using a logging plugin for this – it looks like we're on board to go forward with this solution, so let's I'm going to close #235. Happy to recreate the logger from there as a plugin, if I could help.

ashfurrow commented:
And don't forget about the changelog! :poodle:

esttorhe commented:
Wow :scream: 
Great idea and implementation @aschuch :bow: :ok_hand:

PiXeL16 commented:
Great!

aschuch commented:
Ok, so I've removed the duplication of methods for stubbed/real requests and added the `provider` to the callbacks.

I have also put together a version [on a different branch](https://github.com/aschuch/Moya/commit/34c10bb4644868bc38a64cc84486549cfc556135) that shows how `Plugin` can be implemented in terms of a protocol. This requires a `PluginContainer` wrapper in order to keep a reference within `MoyaProvider` and satisfy the type system. 

I also tried to make  `PluginContainer` conform to `ArrayLiteralConvertible` to keep the API clean, but could not figure out a way to accept the type constraints.

```swift
extension PluginContainer: ArrayLiteralConvertible {
    init<P: Plugin where P.T == T>(arrayLiteral elements: P...) {
        willSendRequestClosures = elements.map { $0.willSendRequest }
        didReceiveResponseClosures = elements.map { $0.didReceiveResponse }
    }
}
```

aschuch commented:
Any opinions towards keeping the subclass vs. the protocol approach?

ashfurrow commented:
Leaning towards protocols, personally. Will have time to review shortly.

ashfurrow commented:
Tricky problem – I think the root of the issue is that we want the plugins to both be able to be used with any type of `MoyaTarget`, but refer to the specific type that's conforming to that target. 

What I mean is, I'm not sure we need to use a typealias'd protocol or a generic class. What about the following?

```swift
public protocol Plugin {

    func willSendRequest<Target: MoyaTarget>(request: Alamofire.Request, provider: MoyaProvider<Target>, token: Target) -> Alamofire.Request
    
    func didReceiveResponse<Target: MoyaTarget>(data: NSData?, statusCode: Int?, response: NSURLResponse?, error: ErrorType?, provider: MoyaProvider<Target>, token: Target)

}
```

Even if someone wanted a plugin to support a _specific_ target type, they should be able to do that:

```swift
class ExamplePlugin: Plugin {

    func willSendRequest<Target: MoyaTarget>(request: Alamofire.Request, provider: MoyaProvider<Target>, token: Target) -> Alamofire.Request {
        return request
    }

    func willSendRequest(request: Alamofire.Request, provider: MoyaProvider<HTTPBin>, token: HTTPBin) -> Alamofire.Request {
        return request
    }

    func didReceiveResponse<Target: MoyaTarget>(data: NSData?, statusCode: Int?, response: NSURLResponse?, error: ErrorType?, provider: MoyaProvider<Target>, token: Target) {

    }
}
```

@aschuch I'll send a PR over to your fork – looking forward to hearing what you think.

ashfurrow commented:
Hmm, wait that might not work. Anyway, I'll play with it.

ashfurrow commented:
Nah, this is silly. You have a working solution, let's merge this PR and keep the idea of moving to a non-inheritance approach for another release.

aschuch commented:
Awesome. Thanks for your input everyone. :heart:
